README.md
# libwebp

This is a binding to [the libwebp library](https://developers.google.com/speed/webp/download).

## Usage

### Preparation

```toml
# Cargo.toml

[dependencies]
libwebp = { version = "0.1.2", features = ["0_6"] }
```

### Simple decoding

```rust
use libwebp::WebPDecodeRGBA;

let data: &[u8];

let (width, height, buf) = WebPDecodeRGBA(data).unwrap();
eprintln!("width = {}, height = {}", width, height);
eprintln!(
    "top-left pixel: rgba({}, {}, {}, {})",
    buf[0],
    buf[1],
    buf[2],
    buf[3] as f64 / 255.0,
)
```

### Simple encoding

```rust
use libwebp::WebPEncodeRGBA;

let buf: &[u8] = &[
    255, 255, 255, 255, // white
    255, 0, 0, 255, // red
    0, 255, 0, 255, // green
    0, 0, 255, 255, // blue
];
let data = WebPEncodeRGBA(buf, 2, 2, 8, 75.0).unwrap();
let lossless_data = WebPEncodeLosslessRGBA(buf, 2, 2, 8).unwrap();
```

## Minimum Supported Rust Version (MSRV)

Rust 1.31.0

## Features

- `demux` ... enables `libwebpdemux` functions.
- `mux` ... enables `libwebpmux` functions.
- `0_5` ... enables functions introduced in libwebp 0.5.0.
- `0_6` ... enables functions introduced in libwebp 0.6.0.
- `1_1` ... enables functions introduced in libwebp 1.1.0.
- `static` ... statically link against the bundled libwebp.
- `extern-types` ... enables `#![feature(extern_types)]`.

## Linking

If libwebp is found in the system, it links against the library.
Otherwise it builds and links against the bundled libwebp.

In these cases, static link is preferred:

- For musl target.
- When cross-compiling.
- `static` feature is turned on.
- `LIBWEBP_SYS_STATIC` environment variable is set to `1` when building.

## Related repositories

- https://github.com/qnighy/libwebp-sys2-rs
- https://github.com/qnighy/libwebp-image-rs

## Completeness

- `types.h`
  - [ ] `WebPMalloc`
  - [x] `WebPFree`
- `decode.h`
  - [x] `WebPGetDecoderVersion`
  - [x] `WebPGetInfo`
  - [x] `WebPDecode*`
  - [x] `WebPDecodeYUV`
  - [x] `WebPDecode*Into`
  - [x] `WebPDecodeYUVInto`
  - [x] `WEBP_CSP_MODE`
  - [x] `WebPIsPremultipliedMode`
  - [x] `WebPIsAlphaMode`
  - [x] `WebPIsRGBMode`
  - [ ] `WebPRGBABuffer`
  - [ ] `WebPYUVABuffer`
  - [ ] `WebPDecBuffer`
  - [ ] `WebPInitDecBuffer`
  - [ ] `WebPFreeDecBuffer`
  - [x] `VP8StatusCode`
  - [x] `WebPIDecoder` (internal memory)
  - [ ] `WebPIDecoder` (external memory)
  - [x] `WebPINewDecoder` (internal memory)
  - [ ] `WebPINewDecoder` (external memory)
  - [x] `WebPINewRGB` (internal memory)
  - [ ] `WebPINewRGB` (external memory)
  - [x] `WebPINewYUVA` (internal memory)
  - [ ] `WebPINewYUVA` (external memory)
  - [x] `WebPIDelete`
  - [x] `WebPIAppend`
  - [ ] `WebPIUpdate`
  - [x] `WebPIDecGetRGB`
  - [x] `WebPIDecGetYUVA`
  - [ ] `WebPIDecodedArea`
  - [ ] `WebPBitstreamFeatures`
  - [ ] `WebPGetFeatures`
  - [ ] `WebPDecoderOptions`
  - [ ] `WebPDecoderConfig`
  - [ ] `WebPInitDecoderConfig`
  - [ ] `WebPIDecode`
  - [ ] `WebPDecode`
- `encode.h`
  - [x] `WebPGetEncoderVersion`
  - [x] `WebPEncode*`
  - [x] `WebPEncodeLossless*`
  - [ ] `WebPImageHint`
  - [ ] `WebPConfig`
  - [ ] `WebPPreset`
  - [ ] `WebPConfigInit`
  - [ ] `WebPConfigPreset`
  - [ ] `WebPConfigLosslessPreset`
  - [ ] `WebPValidateConfig`
  - [ ] `WebPAuxStats`
  - [ ] `WebPWriterFunction`
  - [ ] `WebPMemoryWriter`
  - [ ] `WebPMemoryWriterInit`
  - [ ] `WebPMemoryWriterClear`
  - [ ] `WebPMemoryWrite`
  - [ ] `WebPProgressHook`
  - [ ] `WebPEncCSP`
  - [ ] `WebPEncodingError`
  - [ ] `WEBP_MAX_DIMENSION`
  - [ ] `WebPPicture`
  - [ ] `WebPPictureInit`
  - [ ] `WebPPictureAlloc`
  - [ ] `WebPPictureFree`
  - [ ] `WebPPictureCopy`
  - [ ] `WebPPlaneDistortion`
  - [ ] `WebPPictureDistortion`
  - [ ] `WebPPictureCrop`
  - [ ] `WebPPictureView`
  - [ ] `WebPPictureIsView`
  - [ ] `WebPPictureRescale`
  - [ ] `WebPPictureImportRGB`
  - [ ] `WebPPictureImportRGBA`
  - [ ] `WebPPictureImportRGBX`
  - [ ] `WebPPictureImportBGR`
  - [ ] `WebPPictureImportBGRA`
  - [ ] `WebPPictureImportBGRX`
  - [ ] `WebPPictureARGBToYUVA`
  - [ ] `WebPPictureARGBToYUVADithered`
  - [ ] `WebPPictureSharpARGBToYUVA`
  - [ ] `WebPPictureSmartARGBToYUVA`
  - [ ] `WebPPictureYUVAToARGB`
  - [ ] `WebPCleanupTransparentArea`
  - [ ] `WebPPictureHasTransparency`
  - [ ] `WebPBlendAlpha`
  - [ ] `WebPEncode`
- `mux_types.h`
  - Not at all
- `demux.h`
  - Not at all
- `mux.h`
  - Not at all




CHANGELOG.md
## Unreleased

- MSRV is now 1.33.0
- Internal: bump dev-dependency of rand
- Split `WebPIDecoder` into `WebPIDecoder` and `WebPIDecoderBox`

## 0.1.2

- Implement bindings for incremental decoding
- Refactor: expand macros in `encode.rs`
- Add docs for simple decoding/encoding
- Implement auto traits for `WebpBox` and `WebpYuvBox`

## 0.1.1

- Refactorings:
  - Split out `encode.rs` and `decode.rs`
  - Move `wrap_bytes` to `boxed.rs`
  - Abbreviate `libwebp_sys` as `sys`
- Fix WebPDecoderBGR which mistakenly delegated to RGB
- Implement `WebPDecode*Into`, including `WebPDecodeYUVInto`

## 0.1.0

Initial release.


src/boxed.rs
//! Safe RAII wrappers for `WebPFree`.

use std::fmt;
use std::marker::PhantomData;
use std::mem;
use std::ops::{Deref, DerefMut};
use std::os::raw::*;
use std::panic::{RefUnwindSafe, UnwindSafe};
use std::ptr::NonNull;
use std::slice;

use crate::error::WebPSimpleError;

/// A safe RAII wrapper for `WebPFree`.
///
/// `WebpBox` is much like `Box`, except what function is used for freeing.
///
/// ## Example
///
/// The main usecase is as a return value from libwebp.
///
/// ```rust
/// use libwebp::WebPEncodeRGBA;
///
/// let buf: &[u8];
/// # let buf: &[u8] = &[
/// #     255, 255, 255, 255,
/// #     255, 0, 0, 255,
/// #     0, 255, 0, 255,
/// #     0, 0, 255, 255,
/// # ];
/// // This variable has the type of `WebpBox<[u8]>`.
/// let data = WebPEncodeRGBA(buf, 2, 2, 8, 75.0).unwrap();
/// ```
pub struct WebpBox<T: ?Sized> {
    ptr: NonNull<T>,
    _marker: PhantomData<T>,
}

unsafe impl<T: ?Sized + Send> Send for WebpBox<T> {}
unsafe impl<T: ?Sized + Sync> Sync for WebpBox<T> {}
impl<T: ?Sized + UnwindSafe> UnwindSafe for WebpBox<T> {}
impl<T: ?Sized + RefUnwindSafe> RefUnwindSafe for WebpBox<T> {}

impl<T: ?Sized> WebpBox<T> {
    /// Creates `WebpBox` from a raw pointer.
    ///
    /// ## Safety
    ///
    /// - `raw` must be non-null.
    /// - `raw` must be well-aligned.
    /// - The pointee must be valid as `T`.
    /// - The pointee must be exclusively accessible.
    /// - `raw` must be freeable via `WebPFree`.
    pub unsafe fn from_raw(raw: *mut T) -> WebpBox<T> {
        Self {
            ptr: NonNull::new_unchecked(raw),
            _marker: PhantomData,
        }
    }

    /// Turns `WebpBox` into a raw pointer without freeing anything.
    pub fn into_raw(b: WebpBox<T>) -> *mut T {
        let ptr = b.ptr;
        mem::forget(b);
        ptr.as_ptr()
    }
}

impl<T: ?Sized> Deref for WebpBox<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        unsafe { self.ptr.as_ref() }
    }
}

impl<T: ?Sized> DerefMut for WebpBox<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        unsafe { self.ptr.as_mut() }
    }
}

impl<T: ?Sized> Drop for WebpBox<T> {
    fn drop(&mut self) {
        unsafe {
            WebPFree(self.ptr.as_ptr() as *mut c_void);
        }
    }
}

#[cfg(feature = "0.5")]
use libwebp_sys::WebPFree;

#[cfg(not(feature = "0.5"))]
#[allow(non_snake_case)]
unsafe fn WebPFree(ptr: *mut c_void) {
    extern "C" {
        fn free(ptr: *mut c_void);
    }
    free(ptr);
}

impl<T: fmt::Debug + ?Sized> fmt::Debug for WebpBox<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self as &T, f)
    }
}

#[inline]
pub(crate) unsafe fn wrap_bytes<F>(
    ptr: *mut u8,
    get_len: F,
) -> Result<WebpBox<[u8]>, WebPSimpleError>
where
    F: FnOnce() -> usize,
{
    if !ptr.is_null() {
        let len = get_len();
        Ok(WebpBox::from_raw(slice::from_raw_parts_mut(ptr, len)))
    } else {
        Err(WebPSimpleError)
    }
}

/// A variant of `WebpBox` for return values from `WebPDecodeYUV`.
///
/// See [`WebPDecodeYUV`] for examples.
///
/// [`WebPDecodeYUV`]: ../fn.WebPDecodeYUV.html
pub struct WebpYuvBox {
    y: NonNull<[u8]>,
    u: NonNull<[u8]>,
    v: NonNull<[u8]>,
}

unsafe impl Send for WebpYuvBox {}
unsafe impl Sync for WebpYuvBox {}

impl WebpYuvBox {
    /// Creates `WebpYuvBox` from raw pointers.
    ///
    /// ## Safety
    ///
    /// - `y`, `u`, `v` must be non-null.
    /// - The pointees must be valid as `[u8]`.
    /// - The pointees must be exclusively accessible.
    /// - The head pointer of `y` must be freeable via `WebPFree`.
    /// - The pointees of `u` and `v` must be within the allocated area
    ///   designated by the head pointer of `y`.
    pub unsafe fn from_raw_yuv(y: *mut [u8], u: *mut [u8], v: *mut [u8]) -> WebpYuvBox {
        Self {
            y: NonNull::new_unchecked(y),
            u: NonNull::new_unchecked(u),
            v: NonNull::new_unchecked(v),
        }
    }

    /// Turns `WebpYuvBox` into raw pointers without freeing anything.
    pub fn into_raw_yuv(self) -> (*mut [u8], *mut [u8], *mut [u8]) {
        let y = self.y.as_ptr();
        let u = self.u.as_ptr();
        let v = self.v.as_ptr();
        mem::forget(self);
        (y, u, v)
    }

    /// Immutably deferences to the `y` slice.
    pub fn y(&self) -> &[u8] {
        unsafe { self.y.as_ref() }
    }
    /// Mutably deferences to the `y` slice.
    pub fn y_mut(&mut self) -> &mut [u8] {
        unsafe { self.y.as_mut() }
    }

    /// Immutably deferences to the `u` slice.
    pub fn u(&self) -> &[u8] {
        unsafe { self.u.as_ref() }
    }
    /// Mutably deferences to the `u` slice.
    pub fn u_mut(&mut self) -> &mut [u8] {
        unsafe { self.u.as_mut() }
    }

    /// Immutably deferences to the `v` slice.
    pub fn v(&self) -> &[u8] {
        unsafe { self.v.as_ref() }
    }
    /// Mutably deferences to the `v` slice.
    pub fn v_mut(&mut self) -> &mut [u8] {
        unsafe { self.v.as_mut() }
    }

    /// Immutably, simultaneously dereferences to the `y`, `u`, and `v` slices.
    pub fn yuv(&self) -> (&[u8], &[u8], &[u8]) {
        let y = unsafe { self.y.as_ref() };
        let u = unsafe { self.u.as_ref() };
        let v = unsafe { self.v.as_ref() };
        (y, u, v)
    }
    /// Mutably, simultaneously dereferences to the `y`, `u`, and `v` slices.
    pub fn yuv_mut(&mut self) -> (&mut [u8], &mut [u8], &mut [u8]) {
        let y = unsafe { self.y.as_mut() };
        let u = unsafe { self.u.as_mut() };
        let v = unsafe { self.v.as_mut() };
        (y, u, v)
    }

    /// Turns into a `y` pointer, discarding `u`, `v` slices.
    pub fn into_y(self) -> WebpBox<[u8]> {
        let y = self.y;
        mem::forget(self);
        WebpBox {
            ptr: y,
            _marker: PhantomData,
        }
    }
}

impl Drop for WebpYuvBox {
    fn drop(&mut self) {
        unsafe {
            WebPFree(self.y.as_ptr() as *mut c_void);
        }
    }
}

impl fmt::Debug for WebpYuvBox {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("WebpYuvBox")
            .field("y", &self.y())
            .field("u", &self.u())
            .field("v", &self.v())
            .finish()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use std::panic::{RefUnwindSafe, UnwindSafe};

    #[allow(unused)]
    fn test_auto_traits() {
        fn is_send<T: Send>() {}
        fn is_sync<T: Sync>() {}
        fn is_unwind_safe<T: UnwindSafe>() {}
        fn is_ref_unwind_safe<T: RefUnwindSafe>() {}

        is_send::<WebpBox<[u8]>>();
        is_sync::<WebpBox<[u8]>>();
        is_unwind_safe::<WebpBox<[u8]>>();
        is_ref_unwind_safe::<WebpBox<[u8]>>();

        is_send::<WebpYuvBox>();
        is_sync::<WebpYuvBox>();
        is_unwind_safe::<WebpYuvBox>();
        is_ref_unwind_safe::<WebpYuvBox>();
    }
}


src/encode.rs
use libwebp_sys as sys;
use std::os::raw::*;
use std::ptr;

use crate::boxed::{wrap_bytes, WebpBox};
use crate::error::WebPSimpleError;

/// Return the encoder's version number, packed in hexadecimal using 8bits for
/// each of major/minor/revision.
///
/// E.g: v2.5.7 is 0x020507.
///
/// ## Examples
///
/// ```rust
/// use libwebp::WebPGetEncoderVersion;
///
/// let version = WebPGetEncoderVersion();
/// ```
#[allow(non_snake_case)]
pub fn WebPGetEncoderVersion() -> u32 {
    (unsafe { sys::WebPGetEncoderVersion() }) as u32
}

fn encode_size_check(len: usize, width: u32, height: u32, stride: u32, pixelwidth: usize) {
    assert_eq!(
        width as c_int as u32, width,
        "width {} not within c_int",
        width
    );
    assert_eq!(
        height as c_int as u32, height,
        "height {} not within c_int",
        height
    );
    assert_eq!(
        stride as c_int as u32, stride,
        "stride {} not within c_int",
        stride
    );
    let width = width as usize;
    let height = height as usize;
    let equal = if width == 0 {
        len == 0
    } else {
        len % pixelwidth == 0 && len / pixelwidth % width == 0 && len / pixelwidth / width == height
    };
    assert!(
        equal,
        "buffer size mismatch: {} * {} * {} != {}",
        width, height, pixelwidth, len
    );
}

/// Same as [`WebPEncodeRGBA`], but expecting R, G, B, R, G, B...
/// ordered data.
///
/// [`WebPEncodeRGBA`]: fn.WebPEncodeRGBA.html
#[allow(non_snake_case)]
pub fn WebPEncodeRGB(
    rgb: &[u8],
    width: u32,
    height: u32,
    stride: u32,
    quality_factor: f32,
) -> Result<WebpBox<[u8]>, WebPSimpleError> {
    encode_size_check(rgb.len(), width, height, stride, 3);
    let mut output: *mut u8 = ptr::null_mut();
    let result = unsafe {
        sys::WebPEncodeRGB(
            rgb.as_ptr(),
            width as c_int,
            height as c_int,
            stride as c_int,
            quality_factor as c_float,
            &mut output,
        )
    };
    if result != 0 {
        unsafe { wrap_bytes(output, || result) }
    } else {
        Err(WebPSimpleError)
    }
}

/// Same as [`WebPEncodeRGBA`], but expecting B, G, R, B, G, R...
/// ordered data.
///
/// [`WebPEncodeRGBA`]: fn.WebPEncodeRGBA.html
#[allow(non_snake_case)]
pub fn WebPEncodeBGR(
    bgr: &[u8],
    width: u32,
    height: u32,
    stride: u32,
    quality_factor: f32,
) -> Result<WebpBox<[u8]>, WebPSimpleError> {
    encode_size_check(bgr.len(), width, height, stride, 3);
    let mut output: *mut u8 = ptr::null_mut();
    let result = unsafe {
        sys::WebPEncodeBGR(
            bgr.as_ptr(),
            width as c_int,
            height as c_int,
            stride as c_int,
            quality_factor as c_float,
            &mut output,
        )
    };
    if result != 0 {
        unsafe { wrap_bytes(output, || result) }
    } else {
        Err(WebPSimpleError)
    }
}

/// Encodes images pointed to by `rgba` and returns the WebP binary data.
///
/// The ordering of samples in memory is R, G, B, A, R, G, B, A... in scan
/// order (endian-independent).
///
/// ## Compression
///
/// This function compresses using the lossy format, and the `quality_factor`
/// can go from 0 (smaller output, lower quality) to 100 (best quality,
/// larger output).
///
/// For lossless compression, you can use the [`WebPEncodeLosslessRGBA`]
/// family.
///
/// [`WebPEncodeLosslessRGBA`]: fn.WebPEncodeLosslessRGBA.html
///
/// ## Errors
///
/// Returns `Err` if `data` doesn't contain a valid WebP header.
///
/// ## Panics
///
/// Panics when `stride` is too small or `rgba` has a wrong size.
///
/// ## Variants
///
/// - `WebPEncodeRGBA`
/// - [`WebPEncodeBGRA`]
/// - [`WebPEncodeRGB`]
/// - [`WebPEncodeBGR`]
///
/// [`WebPEncodeBGRA`]: fn.WebPEncodeBGRA.html
/// [`WebPEncodeRGB`]: fn.WebPEncodeRGB.html
/// [`WebPEncodeBGR`]: fn.WebPEncodeBGR.html
///
/// ## Examples
///
/// ```rust
/// use libwebp::WebPEncodeRGBA;
///
/// let buf: &[u8] = &[
///     255, 255, 255, 255, // white
///     255, 0, 0, 255, // red
///     0, 255, 0, 255, // green
///     0, 0, 255, 255, // blue
/// ];
/// let data = WebPEncodeRGBA(buf, 2, 2, 8, 75.0).unwrap();
/// assert_eq!(&data[..4], b"RIFF");
/// assert_eq!(&data[8..12], b"WEBP");
/// ```
#[allow(non_snake_case)]
pub fn WebPEncodeRGBA(
    rgba: &[u8],
    width: u32,
    height: u32,
    stride: u32,
    quality_factor: f32,
) -> Result<WebpBox<[u8]>, WebPSimpleError> {
    encode_size_check(rgba.len(), width, height, stride, 4);
    let mut output: *mut u8 = ptr::null_mut();
    let result = unsafe {
        sys::WebPEncodeRGBA(
            rgba.as_ptr(),
            width as c_int,
            height as c_int,
            stride as c_int,
            quality_factor as c_float,
            &mut output,
        )
    };
    if result != 0 {
        unsafe { wrap_bytes(output, || result) }
    } else {
        Err(WebPSimpleError)
    }
}

/// Same as [`WebPEncodeRGBA`], but expecting B, G, R, A, B, G, R, A...
/// ordered data.
///
/// [`WebPEncodeRGBA`]: fn.WebPEncodeRGBA.html
#[allow(non_snake_case)]
pub fn WebPEncodeBGRA(
    bgra: &[u8],
    width: u32,
    height: u32,
    stride: u32,
    quality_factor: f32,
) -> Result<WebpBox<[u8]>, WebPSimpleError> {
    encode_size_check(bgra.len(), width, height, stride, 4);
    let mut output: *mut u8 = ptr::null_mut();
    let result = unsafe {
        sys::WebPEncodeBGRA(
            bgra.as_ptr(),
            width as c_int,
            height as c_int,
            stride as c_int,
            quality_factor as c_float,
            &mut output,
        )
    };
    if result != 0 {
        unsafe { wrap_bytes(output, || result) }
    } else {
        Err(WebPSimpleError)
    }
}

/// Same as [`WebPEncodeLosslessRGBA`], but expecting R, G, B, R, G, B...
/// ordered data.
///
/// [`WebPEncodeLosslessRGBA`]: fn.WebPEncodeLosslessRGBA.html
#[allow(non_snake_case)]
pub fn WebPEncodeLosslessRGB(
    rgb: &[u8],
    width: u32,
    height: u32,
    stride: u32,
) -> Result<WebpBox<[u8]>, WebPSimpleError> {
    encode_size_check(rgb.len(), width, height, stride, 3);
    let mut output: *mut u8 = ptr::null_mut();
    let result = unsafe {
        sys::WebPEncodeLosslessRGB(
            rgb.as_ptr(),
            width as c_int,
            height as c_int,
            stride as c_int,
            &mut output,
        )
    };
    if result != 0 {
        unsafe { wrap_bytes(output, || result) }
    } else {
        Err(WebPSimpleError)
    }
}

/// Same as [`WebPEncodeLosslessRGBA`], but expecting B, G, R, B, G, R...
/// ordered data.
///
/// [`WebPEncodeLosslessRGBA`]: fn.WebPEncodeLosslessRGBA.html
#[allow(non_snake_case)]
pub fn WebPEncodeLosslessBGR(
    bgr: &[u8],
    width: u32,
    height: u32,
    stride: u32,
) -> Result<WebpBox<[u8]>, WebPSimpleError> {
    encode_size_check(bgr.len(), width, height, stride, 3);
    let mut output: *mut u8 = ptr::null_mut();
    let result = unsafe {
        sys::WebPEncodeLosslessBGR(
            bgr.as_ptr(),
            width as c_int,
            height as c_int,
            stride as c_int,
            &mut output,
        )
    };
    if result != 0 {
        unsafe { wrap_bytes(output, || result) }
    } else {
        Err(WebPSimpleError)
    }
}

/// Encodes images pointed to by `rgba` and returns the WebP binary data.
///
/// The ordering of samples in memory is R, G, B, A, R, G, B, A... in scan
/// order (endian-independent).
///
/// ## Compression
///
/// This function compresses using the lossless format. Files are usually
/// larger than lossy format, but will not suffer any compression loss.
///
/// For lossy compression, you can use the [`WebPEncodeRGBA`] family.
///
/// [`WebPEncodeRGBA`]: fn.WebPEncodeRGBA.html
///
/// Note these functions, like the lossy versions, use the library's default
/// settings. For lossless this means `exact` is disabled. RGB values in
/// transparent areas will be modified to improve compression. To avoid this,
/// use `WebPEncode()` and set `WebPConfig::exact` to `1`.
///
/// (The Rust binding does not yet have the corresponding function.)
///
/// ## Errors
///
/// Returns `Err` if `data` doesn't contain a valid WebP header.
///
/// ## Panics
///
/// Panics when `stride` is too small or `rgba` has a wrong size.
///
/// ## Variants
///
/// - `WebPEncodeLosslessRGBA`
/// - [`WebPEncodeLosslessBGRA`]
/// - [`WebPEncodeLosslessRGB`]
/// - [`WebPEncodeLosslessBGR`]
///
/// [`WebPEncodeLosslessBGRA`]: fn.WebPEncodeLosslessBGRA.html
/// [`WebPEncodeLosslessRGB`]: fn.WebPEncodeLosslessRGB.html
/// [`WebPEncodeLosslessBGR`]: fn.WebPEncodeLosslessBGR.html
///
/// ## Examples
///
/// ```rust
/// use libwebp::WebPEncodeLosslessRGBA;
///
/// let buf: &[u8] = &[
///     255, 255, 255, 255, // white
///     255, 0, 0, 255, // red
///     0, 255, 0, 255, // green
///     0, 0, 255, 255, // blue
/// ];
/// let data = WebPEncodeLosslessRGBA(buf, 2, 2, 8).unwrap();
/// assert_eq!(&data[..4], b"RIFF");
/// assert_eq!(&data[8..12], b"WEBP");
/// ```
#[allow(non_snake_case)]
pub fn WebPEncodeLosslessRGBA(
    rgba: &[u8],
    width: u32,
    height: u32,
    stride: u32,
) -> Result<WebpBox<[u8]>, WebPSimpleError> {
    encode_size_check(rgba.len(), width, height, stride, 4);
    let mut output: *mut u8 = ptr::null_mut();
    let result = unsafe {
        sys::WebPEncodeLosslessRGBA(
            rgba.as_ptr(),
            width as c_int,
            height as c_int,
            stride as c_int,
            &mut output,
        )
    };
    if result != 0 {
        unsafe { wrap_bytes(output, || result) }
    } else {
        Err(WebPSimpleError)
    }
}

/// Same as [`WebPEncodeLosslessRGBA`], but expecting B, G, R, A, B, G, R, A...
/// ordered data.
///
/// [`WebPEncodeLosslessRGBA`]: fn.WebPEncodeLosslessRGBA.html
#[allow(non_snake_case)]
pub fn WebPEncodeLosslessBGRA(
    bgra: &[u8],
    width: u32,
    height: u32,
    stride: u32,
) -> Result<WebpBox<[u8]>, WebPSimpleError> {
    encode_size_check(bgra.len(), width, height, stride, 4);
    let mut output: *mut u8 = ptr::null_mut();
    let result = unsafe {
        sys::WebPEncodeLosslessBGRA(
            bgra.as_ptr(),
            width as c_int,
            height as c_int,
            stride as c_int,
            &mut output,
        )
    };
    if result != 0 {
        unsafe { wrap_bytes(output, || result) }
    } else {
        Err(WebPSimpleError)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::decode::*;

    fn lena() -> Vec<u8> {
        include_bytes!("lena.webp").to_vec()
    }

    #[test]
    #[allow(non_snake_case)]
    fn test_WebPEncodeRGB() {
        let (width, height, buf) = WebPDecodeRGB(&lena()).unwrap();
        assert_eq!(width, 128);
        assert_eq!(height, 128);
        WebPEncodeRGB(&buf, width, height, width, 50.0).unwrap();
    }
}


src/lib.rs
//! # libwebp
//!
//! This is a binding to [the libwebp library](https://developers.google.com/speed/webp/download).
//!
//! ## Usage
//!
//! ### Preparation
//!
//! ```toml
//! # Cargo.toml
//!
//! [dependencies]
//! libwebp = { version = "0.1.2", features = ["0_6"] }
//! ```
//!
//! ### Simple decoding
//!
//! You can use [`WebPDecodeRGBA`] or [`WebPDecodeRGBAInto`] families for
//! simple decoding.
//!
//! [`WebPDecodeRGBA`]: fn.WebPDecodeRGBA.html
//! [`WebPDecodeRGBAInto`]: fn.WebPDecodeRGBAInto.html
//!
//! ```rust
//! use libwebp::WebPDecodeRGBA;
//!
//! let data: &[u8];
//! # let data: &[u8] = include_bytes!("lena.webp");
//!
//! let (width, height, buf) = WebPDecodeRGBA(data).unwrap();
//! # assert_eq!((width, height), (128, 128));
//! assert_eq!(buf.len(), width as usize * height as usize * 4);
//! eprintln!("width = {}, height = {}", width, height);
//! eprintln!(
//!     "top-left pixel: rgba({}, {}, {}, {})",
//!     buf[0],
//!     buf[1],
//!     buf[2],
//!     buf[3] as f64 / 255.0,
//! )
//! ```
//!
//! ### Simple encoding
//!
//! You can use [`WebPEncodeRGBA`] or [`WebPEncodeLosslessRGBA`] families for
//! simple encoding.
//!
//! [`WebPEncodeRGBA`]: fn.WebPEncodeRGBA.html
//! [`WebPEncodeLosslessRGBA`]: fn.WebPEncodeLosslessRGBA.html
//!
//! ```rust
//! use libwebp::{WebPEncodeRGBA, WebPEncodeLosslessRGBA};
//!
//! let buf: &[u8] = &[
//!     255, 255, 255, 255, // white
//!     255, 0, 0, 255, // red
//!     0, 255, 0, 255, // green
//!     0, 0, 255, 255, // blue
//! ];
//! let data = WebPEncodeRGBA(buf, 2, 2, 8, 75.0).unwrap();
//! let lossless_data = WebPEncodeLosslessRGBA(buf, 2, 2, 8).unwrap();
//! assert_eq!(&data[..4], b"RIFF");
//! assert_eq!(&data[8..12], b"WEBP");
//! assert_eq!(&lossless_data[..4], b"RIFF");
//! assert_eq!(&lossless_data[8..12], b"WEBP");
//! ```

pub use crate::decode::*;
pub use crate::encode::*;

pub mod boxed;
mod decode;
mod encode;
pub mod error;
